libdnf/repo/solvable/Package.cpp

```diff
@@ -0,0 +1,235 @@
+    std::shared_ptr<DependencyContainer> getDependencies(Id type, Id marker = -1) const;
+    Queue *getDependencyQueue(Id type, Id marker) const;
+
+    Id id;
+    DnfSack *sack;
+};
+
+libdnf::Package::Impl::Impl(const Impl &source)
+{
+    id  = source.id;
+    sack = source.sack;
+}
+
+void libdnf::Package::Impl::createLibsolvable(HyRepo repo)
+{
+    id = repo_add_solvable(repo->libsolv_repo);
+}
+
+void libdnf::Package::Impl::fillSolvableData(const std::string &name,
+                                              const std::string &version,
+                                              const std::string &arch) const
+{
+    Solvable *libsolvable = pool_id2solvable(dnf_sack_get_pool(sack), id);
+
+    solvable_set_str(libsolvable, SOLVABLE_NAME, name.c_str());
+    solvable_set_str(libsolvable, SOLVABLE_EVR, version.c_str());
+    solvable_set_str(libsolvable, SOLVABLE_ARCH, arch.c_str());
+}
+
+void libdnf::Package::Impl::fillSolvableData(const std::string &name,
+                                              const std::string &version,
+                                              const std::string &arch,
+                                              const std::vector<Dependency> &dependencies) const
+{
+    Solvable *libsolvable = pool_id2solvable(dnf_sack_get_pool(sack), id);
+
+    fillSolvableData(name, version, arch);
+
+    for (const auto &dependency : dependencies) {
+        solvable_add_deparray(libsolvable, dependency.getType(), dependency.getId(), -1);
+    }
+}
+
+std::shared_ptr<DependencyContainer> libdnf::Package::Impl::getDependencies(Id type, Id marker) const
+{
+    auto queue = getDependencyQueue(type, marker);
+    auto container = std::make_shared<DependencyContainer>(sack, *queue);
+
+    queue_free(queue);
+    delete queue;
+
+    return container;
+}
+
+Queue *libdnf::Package::Impl::getDependencyQueue(Id type, Id marker) const
+{
+    Queue dependencyQueue{};
+    auto queue = static_cast<Queue *>(malloc(sizeof(Queue)));
+
+    queue_init(queue);
+    queue_init(&dependencyQueue);
+
+    solvable_lookup_deparray(pool_id2solvable(dnf_sack_get_pool(sack), id), type, &dependencyQueue, marker);
+
+    for (int i = 0; i < dependencyQueue.count; i++) {
+        Id id = dependencyQueue.elements[i];
+        if (id != SOLVABLE_PREREQMARKER)
+            queue_push(queue, id);
+    }
+
+    queue_free(&dependencyQueue);
+
+    return queue;
+}
+
+libdnf::Package::Package(DnfSack *sack, Id id)
+    : pImpl(new Package::Impl())
+{
+    pImpl->sack = sack;
+    pImpl->id = id;
+}
+
+libdnf::Package::Package(DnfSack *sack,
+                           HyRepo repo,
+                           const char *name,
+                           const char *version,
+                           const char *arch)
+        : pImpl(new Package::Impl())
+{
+    pImpl->sack = sack;
+    pImpl->createLibsolvable(repo);
+    pImpl->fillSolvableData(name, version, arch);
```

Comment:
It should fail here - const char to std::string

Reply:
And why it should fail?